server/src/test/java/org/elasticsearch/action/support/nodes/TransportNodesActionTests.java

```diff
@@ -316,6 +318,62 @@ protected Object createActionContext(Task task, TestNodesRequest request) {
         assertTrue(cancellableTask.isCancelled()); // keep task alive
     }
 
+    public void testCompletionAndCancellationShouldMutualExclusivelyHandleResponses() {
+        final var barrier = new CyclicBarrier(2);
+        final var action = new TestTransportNodesAction(
+            clusterService,
+            transportService,
+            new ActionFilters(Set.of()),
+            TestNodeRequest::new,
+            THREAD_POOL.executor(ThreadPool.Names.GENERIC)
+        ) {
+            @Override
+            protected void newResponseAsync(
+                Task task,
+                TestNodesRequest request,
+                Void unused,
+                List<TestNodeResponse> testNodeResponses,
+                List<FailedNodeException> failures,
+                ActionListener<TestNodesResponse> listener
+            ) {
+                final var waited = new AtomicBoolean();
+                for (var response : testNodeResponses) {
+                    if (waited.compareAndSet(false, true)) {
```

Comment:
This is kind of a convoluted way to wait on a nonempty list. There's no concurrency here so the `compareAndSet` is a bit of a sledgehammer. Can we just check `testNodeResponses.isEmpty()`?

Reply:
This is to wait for only the first response. You are right there is no need for AtomicBoolean. I changed it to a primitive boolean variable.